src/replay/replay_recorder.cpp

```diff
@@ -267,13 +462,96 @@ void ReplayRecorder::save()
                     q->m_suspension_length[1],
                     q->m_suspension_length[2],
                     q->m_suspension_length[3],
+                    q->m_skidding_state,
+                    b->m_attachment,
+                    b->m_nitro_amount,
+                    b->m_item_amount,
+                    b->m_item_type,
+                    b->m_special_value,
+                    r->m_distance,
                     r->m_nitro_usage,
                     (int)r->m_zipper_usage,
-                    r->m_skidding_state,
+                    r->m_skidding_effect,
                     (int)r->m_red_skidding,
                     (int)r->m_jumping
                 );
         }   // for i
     }
     fclose(fd);
 }   // save
+
+/* Returns an encoding value for a given attachment type.
+ * The internal values of the enum for attachments may change if attachments
+ * are introduced, removed or even reordered. To avoid compatibility issues
+ * with previous replay files, we add a layer to encode an enum semantical
+ * value the same way, independently of its internal value.
+ * \param type : the type of attachment to encode */
+
+int ReplayRecorder::enumToCode (Attachment::AttachmentType type)
+{
+    int code =
+        (type == Attachment::ATTACH_NOTHING)          ? 0 :
+        (type == Attachment::ATTACH_PARACHUTE)        ? 1 :
+        (type == Attachment::ATTACH_ANVIL)            ? 2 :
+        (type == Attachment::ATTACH_BOMB)             ? 3 :
+        (type == Attachment::ATTACH_SWATTER)          ? 4 :
+        (type == Attachment::ATTACH_BUBBLEGUM_SHIELD) ? 5 :
+                                                       -1 ;
+
+    return code;
+} // enumToCode
+
+/* Returns an encoding value for a given item type
+ * \param type : the type of item to encode */
+
+int ReplayRecorder::enumToCode (PowerupManager::PowerupType type)
+{
+    int code =
```

Comment:
All of that could be replaced with (int)type, no? An enum already has an integer value

Reply:
See my previous comments on the topic in this PR, to you and hiker. This conversion layer is here so that the replay format is not dependent on the enum's internal structure. Encapsulation, in a way : don't rely on the internals. If those enums are changed (new item or attachment added, for example, or assigning new IDs to have easier access on a subset like it has been done for minor race modes), those functions allow replays written before the changes to keep working automatically, instead of having them broken, having to debug the issue when it'll surface out, and then either dropping support for old replays or adding a conversion layer.